### PR TITLE
Refactor to use rcl

### PR DESCRIPTION
**Pull request by jacquelinekay**
Connects to #204

TODO:

- [x] address problems linking to all rmw implementations
- [x] address intra process test failure
- [x] address multithreaded test failure
- [x] services hang (connext and connext dynamic only)

### CODE REVIEW

**Comment by jacquelinekay** [comment]
@wjwwood I think f1625d4 fixes the linking issues we were looking at.

**Comment by jacquelinekay** [comment]
OK, 88e2cd8 fixes the isolated build, though I'm still not sure if I'm following cmake/ament best practices (as usual). I also removed the commit that @wjwwood recently commented on pointing out that an isolated build probably won't work (f1625d4), sorry if that's confusing.

**Comment by jacquelinekay** [comment]
nevermind. the commit does not fix the isolated build. I'm working on it

**Comment by jacquelinekay** [comment]
this is (I think) the most recent CI job for this branch:

http://ci.ros2.org/job/ci_osx/934/

As you can see, the services test for Connext and Connext Dynamic times out because services hang indefinitely.

The failures in test_communication should be fixed by https://github.com/ros2/system_tests/pull/113 and are unrelated to the underlying problem of services hanging indefinitely.

**Comment by jacquelinekay** [comment]
With some fixes to rcl services, I'm relaunching CI

http://ci.ros2.org/job/ci_linux/1195/
http://ci.ros2.org/job/ci_osx/971/
http://ci.ros2.org/job/ci_windows/1250/

File: rclcpp/CMakeLists.txt [hunk 1]
```diff
@@ -90,6 +90,7 @@ if(AMENT_ENABLE_TESTING)
   ament_add_gtest(test_mapped_ring_buffer test/test_mapped_ring_buffer.cpp)
   if(TARGET test_mapped_ring_buffer)
     target_include_directories(test_mapped_ring_buffer PUBLIC
+      ${rcl_INCLUDE_DIRS}
       ${rcl_interfaces_INCLUDE_DIRS}
       ${rmw_INCLUDE_DIRS}
```
Comment (wjwwood) [inline rclcpp/CMakeLists.txt]: Is `rclcpp` still using `rmw` in places? Can/should references to `rmw` be removed from the `rclcpp` build files?

File: rclcpp/cmake/get_rclcpp_information.cmake [hunk 2]
```diff
@@ -62,9 +67,7 @@ macro(get_rclcpp_information rmw_implementation var_prefix)
 
   # dependencies
   set(_exported_dependencies
-    "rcl_interfaces"
-    "rmw"
-    "${rmw_implementation}"
+    "rcl${target_suffix}"
```
Comment (wjwwood) [inline rclcpp/cmake/get_rclcpp_information.cmake]: Should `rcl_interfaces` be exported here? It was removed.

File: rclcpp/include/rclcpp/client.hpp [hunk 3]
```diff
@@ -23,9 +23,13 @@
 #include <tuple>
 #include <utility>
 
+#include "rcl/error_handling.h"
+#include "rcl/client.h"
```
Comment (wjwwood) [inline rclcpp/include/rclcpp/client.hpp]: nitpick: alpha-ordering

File: rclcpp/include/rclcpp/client.hpp [hunk 4]
```diff
@@ -23,9 +23,13 @@
...
+
 #include "rclcpp/function_traits.hpp"
 #include "rclcpp/macros.hpp"
 #include "rclcpp/utilities.hpp"
+#include "rclcpp/type_support_decl.hpp"
```
Comment (wjwwood) [inline rclcpp/include/rclcpp/client.hpp]: alpha

File: rclcpp/include/rclcpp/client.hpp [hunk 5]
```diff
@@ -42,32 +46,28 @@ class ClientBase
 
   RCLCPP_PUBLIC
   ClientBase(
-    std::shared_ptr<rmw_node_t> node_handle,
-    rmw_client_t * client_handle,
+    std::shared_ptr<rcl_node_t> node_handle,
     const std::string & service_name);
 
   RCLCPP_PUBLIC
-  virtual ~ClientBase();
```
Comment (wjwwood) [inline rclcpp/include/rclcpp/client.hpp]: Why did you remove the destructor declaration? I think it might be important that it is declared `virtual`.

File: rclcpp/include/rclcpp/client.hpp [hunk 6]
```diff
@@ -42,32 +46,28 @@ class ClientBase
...
+  const rcl_client_t *
   get_client_handle() const;
 
   virtual std::shared_ptr<void> create_response() = 0;
   virtual std::shared_ptr<rmw_request_id_t> create_request_header() = 0;
   virtual void handle_response(
     std::shared_ptr<rmw_request_id_t> request_header, std::shared_ptr<void> response) = 0;
 
-private:
+protected:
```
Comment (wjwwood) [inline rclcpp/include/rclcpp/client.hpp]: Any rational here? making members protected is possibly the right thing to do, but it's not clear why you decided to do this.

File: rclcpp/include/rclcpp/client.hpp [hunk 7]
```diff
@@ -93,11 +93,32 @@ class Client : public ClientBase
...
 
   Client(
-    std::shared_ptr<rmw_node_t> node_handle,
-    rmw_client_t * client_handle,
-    const std::string & service_name)
-  : ClientBase(node_handle, client_handle, service_name)
-  {}
+    std::shared_ptr<rcl_node_t> node_handle,
+    const std::string & service_name,
+    rcl_client_options_t & client_options)
```
Comment (wjwwood) [inline rclcpp/include/rclcpp/client.hpp]: You changed the order of the arguments here, does this not need changes in the examples and stuff like that?

Comment (wjwwood) [inline rclcpp/include/rclcpp/client.hpp]: nvm, this is not used directly by the user.

File: rclcpp/include/rclcpp/executor.hpp [hunk 8]
```diff
@@ -292,10 +295,6 @@ class Executor
   get_next_timer(AnyExecutable::SharedPtr any_exec);
 
   RCLCPP_PUBLIC
-  std::chrono::nanoseconds
-  get_earliest_timer();
```
Comment (wjwwood) [inline rclcpp/include/rclcpp/executor.hpp]: Why was this removed?

File: rclcpp/include/rclcpp/client.hpp (hunk 5 above)
Comment (jacquelinekay) [inline rclcpp/include/rclcpp/client.hpp]: I'm not sure. Previously the child templated class `Client` did not implement a custom destructor, and I add one in this pull request. Maybe I needed to delete this declaration to make it work. I can put this back in and see.

File: rclcpp/include/rclcpp/executor.hpp (hunk 8 above)
Comment (wjwwood) [inline rclcpp/include/rclcpp/executor.hpp]: nvm, I guess it's probably handled by `rcl_wait` now. Sorry for the noise.

File: rclcpp/include/rclcpp/client.hpp (hunk 6 above)
Comment (jacquelinekay) [inline rclcpp/include/rclcpp/client.hpp]: I decided to push some member variables that are not templated but need to be accessible by the child class into the parent class.

File: rclcpp/cmake/get_rclcpp_information.cmake (hunk 2 above)
Comment (jacquelinekay) [inline rclcpp/cmake/get_rclcpp_information.cmake]: I can add it back in, not sure why I removed it.

File: rclcpp/include/rclcpp/node_impl.hpp [hunk 9]
```diff
@@ -314,27 +295,17 @@ Node::create_client(
...
-  if (!client_handle) {
-    // *INDENT-OFF* (prevent uncrustify from making unecessary indents here)
-    throw std::runtime_error(
-      std::string("could not create client: ") +
-      rmw_get_error_string_safe());
-    // *INDENT-ON*
-  }
+  rcl_client_options_t options = rcl_client_get_default_options();
+  options.qos = qos_profile;
+  // options.allocator = rclcpp::allocator::get_rcl_allocator<ServiceT, Alloc>(*allocator.get());
```
Comment (wjwwood) [inline rclcpp/include/rclcpp/node_impl.hpp]: uncomment or remove?

File: rclcpp/include/rclcpp/client.hpp (hunk 5 above)
Comment (jacquelinekay) [inline rclcpp/include/rclcpp/client.hpp]: When I add it back I get:
```
librclcpp__rmw_opensplice_cpp.so: undefined reference to `rclcpp::client::ClientBase::~ClientBase()'
librclcpp__rmw_opensplice_cpp.so: undefined reference to `typeinfo for rclcpp::client::ClientBase'
librclcpp__rmw_opensplice_cpp.so: undefined reference to `vtable for rclcpp::client::ClientBase'
```

Comment (wjwwood) [inline rclcpp/include/rclcpp/client.hpp]: It needs to be implemented or made pure virtual. I'd say give it an empty implementation in the `.cpp` file.

File: rclcpp/include/rclcpp/publisher.hpp [hunk 10]
```diff
@@ -152,15 +151,54 @@ class Publisher : public PublisherBase
...
+    }
+    if (rmw_get_gid_for_publisher(publisher_rmw_handle, &rmw_gid_) != RMW_RET_OK) {
+      // *INDENT-OFF* (prevent uncrustify from making unecessary indents here)
+      throw std::runtime_error(
+        std::string("failed to get publisher gid: ") + rmw_get_error_string_safe());
+      // *INDENT-ON*
+    }
+  }
+
+  ~Publisher()
```
Comment (wjwwood) [inline rclcpp/include/rclcpp/publisher.hpp]: Please mark this as `virtual`.

File: rclcpp/include/rclcpp/service.hpp [hunk 11]
```diff
@@ -39,19 +43,15 @@ class ServiceBase
 
   RCLCPP_PUBLIC
   ServiceBase(
-    std::shared_ptr<rmw_node_t> node_handle,
-    rmw_service_t * service_handle,
+    std::shared_ptr<rcl_node_t> node_handle,
     const std::string service_name);
 
   RCLCPP_PUBLIC
-  virtual ~ServiceBase();
```
Comment (wjwwood) [inline rclcpp/include/rclcpp/service.hpp]: Same thing here, I think you should implement it rather than remove it.

File: rclcpp/include/rclcpp/publisher.hpp (hunk 10 above)
Comment (jacquelinekay) [inline rclcpp/include/rclcpp/publisher.hpp]: But this is the child class, not the parent class. And it's templated so it can't be inherited from.

File: rclcpp/include/rclcpp/service.hpp (hunk 11 above)
Comment (wjwwood) [inline rclcpp/include/rclcpp/service.hpp]: Without this the child class's destructor will not be called when deleting a pointer of this type (which I think could happen in the future if it doesn't already): http://stackoverflow.com/a/461224/671658

File: rclcpp/include/rclcpp/service.hpp [hunk 12]
```diff
@@ -88,15 +88,37 @@ class Service : public ServiceBase
...
+    {
+      throw std::runtime_error(
+              std::string("could not create service: ") +
+              rcl_get_error_string_safe());
+    }
+  }
 
   Service() = delete;
 
+  ~Service()
```
Comment (wjwwood) [inline rclcpp/include/rclcpp/service.hpp]: This should be `virtual` for the same reason, but in the case where the user makes a child class which inherits from this one.

File: rclcpp/include/rclcpp/publisher.hpp (hunk 10 above)
Comment (jacquelinekay) [inline rclcpp/include/rclcpp/publisher.hpp]: Hmm, nevermind, I guess I'm misremembering how templates and inheritance interact in C++

Comment (wjwwood) [inline rclcpp/include/rclcpp/publisher.hpp]: I think you can inherit from a template class. My understanding is that you should almost always declare a destructor virtual, just in case someone later wants to inherit from it. I believe it applies to templates as well. According to this you can have a template or non-template class inherit from another template class:

http://www.prenhall.com/divisions/esm/app/kafura/secure/chapter7/html/7.5_inheritance.htm

Comment (jacquelinekay) [inline rclcpp/include/rclcpp/publisher.hpp]: ok so, the general pattern I'll follow is: parent class with empty non-virtual destructor implementation, child class with destructor declared as virtual

Comment (wjwwood) [inline rclcpp/include/rclcpp/publisher.hpp]: Yes.

Also if the parent class is already abstract, then you could make the destructor pure virtual and avoid the empty implementation. (which does not appear to be the case here)

**Comment by jacquelinekay** [comment]
welp. Lots of regressions since the last time I tested this. Looking into it

**Comment by wjwwood** [comment]
Well the code review looks ok to me (other than comments). I'll try to help you look into the regressions.

**Comment by jacquelinekay** [comment]
One last round for good measure
http://ci.ros2.org/job/ci_linux/1207/
http://ci.ros2.org/job/ci_osx/976/
http://ci.ros2.org/job/ci_windows/1254

**Comment by jacquelinekay** [comment]
I had to ifdef out the custom allocators in rcl because the way I did it was causing every Windows test to segfault. I also I had to fix some problems with inheritance in timers on Windows.

The multithreaded tests don't pass on Windows on my vm. I'm going to run CI to see what the status is but I'm tempted to just merge this so that I can unblock @wjwwood and then I can work on fixing up the technical debt in parallel.

http://ci.ros2.org/job/ci_linux/1214/
http://ci.ros2.org/job/ci_osx/981/
http://ci.ros2.org/job/ci_windows/1259/

**Comment by jacquelinekay** [comment]
Another round with the atomics fix for Windows (this fixes the multithreaded tests locally for me): https://github.com/ros2/rcl/pull/49

http://ci.ros2.org/job/ci_windows/1264/

**Comment by jacquelinekay** [comment]
OMG that last CI job has 18 failures, which is as many as the nightly Windows release job!! woohoo

**Comment by wjwwood** [comment]
Are they the same ones :)?

**Comment by wjwwood** [comment]
The changes lgtm. Anything else need to be done to merge this?

File: rclcpp/src/rclcpp/utilities.cpp [hunk 13]
```diff
@@ -89,14 +93,11 @@ signal_handler(int signal_value)
 void
 rclcpp::utilities::init(int argc, char * argv[])
 {
-  (void)argc;
-  (void)argv;
   g_is_interrupted.store(false);
-  rmw_ret_t status = rmw_init();
```
Comment (dirk-thomas) [inline rclcpp/src/rclcpp/utilities.cpp]: The call to `rmw_init` was removed but not replaces with any equivalent. As a result the rmw init function is not being called anymore. See discussion on the mailing list: https://groups.google.com/forum/#!topic/ros-sig-ng-ros/jr6x3vWy4ls

Comment (wjwwood) [inline rclcpp/src/rclcpp/utilities.cpp]: I believe `rcl_init` is being called one line below. I believe the problem is that `rcl_init` does not call `rmw_init`, not that `rclcpp` does not call `rmw_init`.

Comment (wjwwood) [inline rclcpp/src/rclcpp/utilities.cpp]: `rcl_init` does not call `rmw_init`, I'll open a pull request for this:

https://github.com/ros2/rcl/pull/69

However, there is no `rmw_shutdown` right now and `rmw_init` is generally not used, so I think I probably didn't use it because I knew we had to change some things about how init/shutdown worked with the rmw interface anyways. I'm not sure it should be merged in it's asymmetric state, or if we should address this when we fix the rmw init/shutdown code.
